### PR TITLE
neuron: add new versions and clean-up the recipe

### PR DESCRIPTION
**Pull request by pramodk**
* update tarball urls, add new versions and update maintainers
* remove unnecessary variant like cross-compile
* use self.define and self.define_from_variant
* other minor refactoring/cleanup

X-ref BlueBrain/spack/pull/2136

### CODE REVIEW

**Review by iomaganaris — APPROVED** [review]
LGTM

**Review by alexsavulescu — APPROVED** [review]

File: var/spack/repos/builtin/packages/neuron/package.py [hunk 1]
```diff
@@ -16,89 +16,139 @@ class Neuron(CMakePackage):
...
     depends_on("py-pytest", when="+python+tests")
     depends_on("py-mpi4py", when="+mpi+python+tests")
-    depends_on("readline")
+    depends_on("py-numpy", when="+python")
+    depends_on("py-cython", when="+rx3d", type="build")
+    depends_on("py-pytest-cov", when="+tests")
+
+    # next two needed after neuronsimulator/nrn#2235.
+    depends_on("py-pip", type=("build"))
+    depends_on("py-packaging", type=("run"))
```
Comment (matz-e) [inline var/spack/repos/builtin/packages/neuron/package.py]:
```suggestion
    depends_on("py-pip", type=("build"))
    depends_on("py-setuptools", type=("build"))
    depends_on("py-packaging", type=("run"))
```

File: var/spack/repos/builtin/packages/neuron/package.py [hunk 2]
```diff
@@ -16,89 +16,140 @@ class Neuron(CMakePackage):
...
     depends_on("py-mpi4py", when="+mpi+python+tests")
-    depends_on("readline")
+    depends_on("py-numpy", when="+python")
+    depends_on("py-cython", when="+rx3d", type="build")
+    depends_on("py-pytest-cov", when="+tests")
+
+    # next two needed after neuronsimulator/nrn#2235.
+    depends_on("py-pip", type=("build"))
+    depends_on("py-setuptools", type=("build"))
+    depends_on("py-packaging", type=("run"))
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/neuron/package.py]:
```suggestion
    depends_on("py-pip", type="build")
    depends_on("py-setuptools", type="build")
    depends_on("py-packaging", type="run")
```

File: var/spack/repos/builtin/packages/neuron/package.py [hunk 3]
```diff
@@ -16,89 +16,140 @@ class Neuron(CMakePackage):
...
+
     depends_on("caliper", when="+caliper")
-    depends_on("py-numpy", type="run")
 
-    conflicts("+rx3d", when="~python")
+    gpu_compiler_message = "neuron: for gpu build use %nvhpc"
+    conflicts("%gcc", when="+gpu", msg=gpu_compiler_message)
+    conflicts("%intel", when="+gpu", msg=gpu_compiler_message)
+    conflicts("%aocc", when="+gpu", msg=gpu_compiler_message)
+    conflicts("%clang", when="+gpu", msg=gpu_compiler_message)
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/neuron/package.py]:
```suggestion
    gpu_compiler_message = "neuron+gpu needs %nvhpc"
    requires("%nvhpc", when="+gpu", msg=gpu_compiler_message)
```

File: var/spack/repos/builtin/packages/neuron/package.py [hunk 4]
```diff
@@ -16,89 +16,140 @@ class Neuron(CMakePackage):
...
+                self.define_from_variant("CORENRN_ENABLE_UNIT_TESTS", "tests"),
+            ]
+
+            nmodl_options = "codegen --force"
+            if spec.satisfies("+sympy"):
+                nmodl_options += " sympy --analytic"
+            options.append(self.define("CORENRN_NMODL_FLAGS", nmodl_options))
+
+            if spec.satisfies("+gpu"):
+                nvcc = which("nvcc")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/neuron/package.py]: Can't we use either `spack_cc` or `self.compiler.cc` (depending on whether you need the wrapper), and avoid a call to `which`?

File: var/spack/repos/builtin/packages/neuron/package.py [hunk 5]
```diff
@@ -16,89 +16,140 @@ class Neuron(CMakePackage):
...
+                options.append(self.define("CMAKE_CUDA_COMPILER", nvcc))
+                options.append(self.define("CORENRN_ENABLE_GPU", True))
+
+            args.extend(options)
+
+        # Enable math optimisations to enable SIMD/vectorisation in release modes
+        if spec.variants["build_type"].value in ["Release", "RelWithDebInfo"]:
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/neuron/package.py]:
```suggestion
        if spec.satisfies("build_type=Release") or spec.satisfies("build_type=RelWithDebInfo"):
```

**Review by alalazo — CHANGES_REQUESTED** [review]
A few minor requests

File: var/spack/repos/builtin/packages/neuron/package.py (hunk 4 above)
Comment (pramodk) [reply var/spack/repos/builtin/packages/neuron/package.py]: those will be `nvc` and not `nvcc`, right? I see that we have a dependency with cuda. So I replaced it with `spec["cuda"].prefix.bin.nvcc`.

**Comment by pramodk** [comment]
thanks @matz-e and @alalazo for the reviews! I have updated the recipe to address the reviews!

**Review by alalazo — APPROVED** [review]

**Pull request merged**